modules/openxAnalyticsAdapter.js

```diff
@@ -634,7 +667,405 @@ adapterManager.registerAnalyticsAdapter({
+    deviceOSType: detectOS(),
+    browser: detectBrowser(),
+    testCode: analyticsConfig.testCode,
+    bidRequests: buildBidRequestsPayload(adUnitCodeToBidderRequestMap),
+  };
+
+  function buildBidRequestsPayload(adUnitCodeToBidderRequestMap) {
+    return utils._map(adUnitCodeToBidderRequestMap, (bidderRequestMap, adUnitCode) => {
+      return utils._map(bidderRequestMap, (bidderRequest) => {
+        let {bidder, source, bids, mediaTypes, timedOut} = bidderRequest;
+        return {
+          adUnitCode,
+          bidder,
+          source,
+          hasBidderResponded: Object.keys(bids).length > 0,
+          availableAdSizes: getMediaTypeSizes(mediaTypes),
+          availableMediaTypes: getMediaTypes(mediaTypes),
+          timedOut,
+          bidResponses: utils._map(bidderRequest.bids, (bidderBidResponse) => {
+            let {
+              cpm,
+              creativeId,
+              responseTimestamp,
+              ts,
+              adId,
+              meta,
+              mediaType,
+              dealId,
+              ttl,
+              netRevenue,
+              currency,
+              originalCpm,
+              originalCurrency,
+              width,
+              height,
+              latency,
+              winner,
+              rendered,
+              renderTime,
+            } = bidderBidResponse;
+
+            return {
+              microCPM: cpm * 1000,
+              netRevenue,
+              currency,
+              mediaType,
+              height,
+              width,
+              size: `${width}x${height}`,
+              dealId,
+              latency,
+              ttl,
+              winner,
+              creativeId,
+              ts,
+              rendered,
+              renderTime,
+              meta,
+            }
+          })
+        }
+      })
+    }).flat();
+  }
+
+  function getMediaTypeSizes(mediaTypes) {
+    return utils._map(mediaTypes, (mediaTypeConfig, mediaType) => {
+      return utils.parseSizesInput(mediaTypeConfig.sizes)
+        .map(size => `${mediaType}_${size}`);
+    }).flat();
+  }
+
+  function getMediaTypes(mediaTypes) {
+    return utils._map(mediaTypes, (mediaTypeConfig, mediaType) => mediaType);
+  }
+}
+
+function getAdUnitByAuctionAndAdId(auction, adId) {
+  let adunit;
+
+  utils._each(auction.adUnitCodeToBidderRequestMap, (bidderRequestIdMap) => {
+    utils._each(bidderRequestIdMap, bidderRequest => {
+      utils._each(bidderRequest.bids, (bid, bidId) => {
+        if (bidId === adId) {
+          adunit = bid;
+        }
+      });
+    });
+  });
+
+  return adunit;
+}
+
 export default Object.assign({
   adapter: openxAdapter,
   auctionEndWaitTime: AUCTION_END_WAIT_TIME
 });
+
+/**
+ * Test Helper Functions
+ */
+
+// reset the cache for unit tests
+openxAdapter.reset = function() {
+  auctionMap = {};
+  auctionOrder = 1;
+};
+
+/**
+ *  Type Definitions
+ */
+
+/**
+ * @typedef {Object} BidResponse
```

Comment:
ohh this is what you meant... jsdoc?

Reply:
yeah...  my ide can read it and use that as type defs for auto suggest and error reporting